packages/account/src/BaseAccount.ts

```diff
@@ -40,99 +23,122 @@ export abstract class SmartAccount implements ISmartAccount {
   }
 
-  isProviderDefine(): boolean {
+  isProviderDefined(): boolean {
     if (!this.provider) throw new Error('Provider is undefined')
 
     return true
   }
 
-  async estimateUserOpGas(userOp: UserOperation): Promise<UserOperation> {
+  async estimateUserOpGas(userOp: Partial<UserOperation>, overrides?: Overrides): Promise<Partial<UserOperation>> {
     const requiredFields: UserOperationKey[] = [
       'sender',
       'nonce',
       'initCode',
-      'callData',
-      'callGasLimit',
-      'paymasterAndData'
+      'callData'
     ]
     this.validateUserOp(userOp, requiredFields)
+
+    // Override gas values in userOp if provided in overrides params
+    if (overrides) {
+      userOp = { ...userOp, ...overrides }
+    }
+
+    console.log('userOp in estimation', userOp);
+
+    // Defining the keys that are related that can be overrides
+    const overrideGasFields: UserOperationKey[] = [
+      'maxFeePerGas',
+      'maxPriorityFeePerGas',
+      'verificationGasLimit',
+      'callGasLimit',
+      'preVerificationGas'
+    ];
+
+   // here we are verifying either all necessary gas properties are present in userOp.
+    let skipEstimations = true;
+    for (const key of overrideGasFields) {
+      if (!userOp[key]) {
+        skipEstimations = false;
+        break;
+      }
+    }
+   // If all necessary properties are present in userOp. we will skip estimation and return userOp
+    if (skipEstimations) {
+      return userOp
+    }
+
     if (!this.bundler) {
       if (!this.provider) throw new Error('Provider is not present for making rpc calls')
-      // if no bundler url is provided run offchain logic to assign following values of user
+      // if no bundler url is provided run offchain logic to assign following values of UserOp
       // maxFeePerGas, maxPriorityFeePerGas, verificationGasLimit, callGasLimit, preVerificationGas
       const feeData = await this.provider.getFeeData()
-      if (EIP1559_UNSUPPORTED_NETWORKS.includes(this.chainId)) {
-        // assign gasPrice to both maxFeePerGas and maxPriorityFeePerGas in case chain does not support Type2 transaction
-        userOp.maxFeePerGas = userOp.maxPriorityFeePerGas =
-          feeData.gasPrice ?? (await this.provider.getGasPrice())
-      } else {
-        if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas)
-          throw new Error('Unable to get maxFeePerGas and maxPriorityFeePerGas from provider')
-        userOp.maxFeePerGas = feeData.maxFeePerGas
-        userOp.maxPriorityFeePerGas = feeData.maxPriorityFeePerGas
-      }
-      userOp.verificationGasLimit = await this.getVerificationGasLimit(userOp.initCode)
-      userOp.callGasLimit = await this.provider.estimateGas({
+      userOp.maxFeePerGas = userOp.maxFeePerGas ?? feeData.maxFeePerGas ?? (await this.provider.getGasPrice())
+      userOp.maxPriorityFeePerGas = userOp.maxPriorityFeePerGas ?? feeData.maxPriorityFeePerGas ?? (await this.provider.getGasPrice())
+      if (userOp.initCode)
```

Comment:
if you see function start it checks for required fields ('sender', 'nonce', 'initCode', 'callData') in the UserOperation object. If any of the field is missing, an exception is thrown. The check if (userOp.initCode) is added because changing userOp type to Partial<UserOperation>.